File: examples/full-cdk-policy/full-cdk-policy.py
import json
import iam_floyd as statement


def get_policy():
    # doc-start
    policy = {
        'Version': '2012-10-17',
        'Statement': [
            # allow all CFN actions
            statement.Cloudformation() \
            .allow() \
            .all_actions() \
            .to_json(),
            # allow access to the CDK staging bucket
            statement.All() \
            .allow() \
            .all_actions() \
            .if_aws_called_via('cloudformation.amazonaws.com') \
            .to_json(),
            # allow access to the CDK staging bucket
            statement.S3() \
            .allow() \
            .all_actions() \
            .on('arn:aws:s3:::cdktoolkit-stagingbucket-*') \
            .to_json(),
            # even when triggered via CFN, do not allow modifications of the
            #  account
            statement.Account() \
            .deny() \
            .all_permission_management_actions() \
            .all_write_actions() \
            .to_json(),
            # even when triggered via CFN, do not allow modifications of the
            #  organization
            statement.Organizations() \
            .deny() \
            .all_permission_management_actions() \
            .all_write_actions() \
            .to_json()
        ]
    }

    # doc-end
    return policy


pretty = json.dumps(get_policy(), indent=4, sort_keys=True)
print(pretty)

File: examples/full-cdk-policy/full-cdk-policy.ts
import { Statement } from '../../lib';

function getPolicy() {
  function wrap() {
    // doc-start
    const policy = {
      Version: '2012-10-17',
      Statement: [
        new Statement.Cloudformation() // allow all CFN actions
          .allow()
          .allActions(),
        new Statement.All() // allow absolutely everything that is triggered via CFN
          .allow()
          .allActions()
          .ifAwsCalledVia('cloudformation.amazonaws.com'),
        new Statement.S3() // allow access to the CDK staging bucket
          .allow()
          .allActions()
          .on('arn:aws:s3:::cdktoolkit-stagingbucket-*'),
        new Statement.Account() // even when triggered via CFN, do not allow modifications of the account
          .deny()
          .allPermissionManagementActions()
          .allWriteActions(),
        new Statement.Organizations() // even when triggered via CFN, do not allow modifications of the organization
          .deny()
          .allPermissionManagementActions()
          .allWriteActions(),
      ],
    };
    // doc-end
    return policy;
  }
  return wrap();
}

const policy = getPolicy();
const str = JSON.stringify(policy, null, 4);
console.log(str);
